android/app/src/main/kotlin/com/nt4f04und/sweyer/MusicPlayerAppWidget.kt

```diff
@@ -0,0 +1,318 @@
+package com.nt4f04und.sweyer
+
+import android.app.PendingIntent
+import android.appwidget.AppWidgetManager
+import android.appwidget.AppWidgetProviderInfo
+import android.content.Context
+import android.content.Intent
+import android.content.SharedPreferences
+import android.content.res.Resources
+import android.graphics.*
+import android.net.Uri
+import android.os.Build
+import android.os.Bundle
+import android.os.CancellationSignal
+import android.os.OperationCanceledException
+import android.util.Log
+import android.util.Size
+import android.util.TypedValue
+import android.view.KeyEvent
+import android.view.View.GONE
+import android.view.View.VISIBLE
+import android.widget.RemoteViews
+import es.antonborri.home_widget.*
+import java.io.IOException
+import java.net.URISyntaxException
+import java.util.concurrent.Executors
+import kotlin.math.max
+import kotlin.math.roundToInt
+import com.nt4f04und.sweyer.sweyer_plugin.Constants
+
+/**
+ * Sweyer music App Widget.
+ * The widget is freely resizable and has a playback control button bar at the bottom.
+ * Depending on the horizontal size available, there is eiter just a play/pause button,
+ * a play/pause and skip button, or a play/pause, skip and previous button.
+ * The background is the current song cover if available. The edges of the widget are rounded.
+ */
+class MusicPlayerAppWidget : HomeWidgetProvider() {
+    override fun onUpdate(
+        context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray, widgetData: SharedPreferences
+    ) = updateWidgets(context, WidgetInfoProvider(context, appWidgetManager, appWidgetIds), widgetData)
+
+    override fun onAppWidgetOptionsChanged(
+        context: Context?, appWidgetManager: AppWidgetManager?, appWidgetId: Int, newOptions: Bundle?
+    ) {
+        super.onAppWidgetOptionsChanged(context, appWidgetManager, appWidgetId, newOptions)
+        if (context == null || appWidgetManager == null) {
+            return
+        }
+        updateWidgets(
+            context,
+            WidgetInfoProvider(context, appWidgetManager, intArrayOf(appWidgetId), newOptions),
+            HomeWidgetPlugin.getData(context),
+        )
+    }
+
+    /**
+     * Update all widgets to for the current playback state and their widget size.
+     * [widgetData] contains the current play state and currently playing song.
+     */
+    private fun updateWidgets(context: Context, widgetInfoProvider: WidgetInfoProvider, widgetData: SharedPreferences) {
+        val playing = widgetData.getBoolean("playing", false)
+        val songUri = try {
+            widgetData.getString("song", null)?.let { uri -> Uri.parse(uri) }
+        } catch (ignored: URISyntaxException) {
+            null
+        }
+        if (songUri == null || Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
```

Comment:
> Is this version constraint because of loadThumbnail?

Yeah. We have the same version constraint in `loadAlbumArt`. I agree it would be to find a way for pre 29 devices to load the thumbnail, according to https://apilevels.com there is still ~9% of all devices that wouldn't have proper images.

Reply:
> We have the same version constraint in `loadAlbumArt`

yeah,  but only because on below Q we use file paths directly

Reply:
maybe there is a support library out there

Reply:
> yeah, but only because on below Q we use file paths directly

Right, I'll look at this when I convert it to Glance in a follow-up PR.